docroot/modules/custom/mass_alerts/src/Plugin/Block/AlertsBlock.php

```diff
@@ -88,21 +89,32 @@ public function build() {
 
     $config = $this->getConfiguration();
     $path = FALSE;
+    $node = $this->routeMatch->getParameter('node');
+    $node_type = $node ? $node->bundle() : '';
 
     if ($config['alerts_block_type'] == 'sitewide') {
       $path = Url::fromRoute('mass_alerts.site_alerts')->toString();
     }
-    else {
-      $node = $this->routeMatch->getParameter('node');
-      if ($node instanceof NodeInterface) {
-        $nid = $node->id();
-        $path = Url::fromRoute('mass_alerts.page_alerts', ['nid' => $nid])->toString();
-      }
+    elseif ($node) {
+      $nid = $node->id();
+      $path = Url::fromRoute('mass_alerts.page_alerts', ['nid' => $nid])->toString();
     }
 
+    // Data to fetch the alerts as soon as possible.
+    // Needs to be cached, as rendering a block happens once.
+    // @see docroot/themes/custom/mass_theme/templates/layout/html.html.twig */
+    // @see mass_theme_preprocess_html()
+    $current_path = \Drupal::service('path.current')->getPath();
+    $cid = 'mass_alerts_blocks:' . $current_path;
+    $cache_entry = \Drupal::cache('render')->get($cid);
+    $data = $cache_entry->data ?? new stdClass();
+    $data->$path = ['#path' => $path];
+    \Drupal::cache('render')->set($cid, $data);
+
```

Comment:
Can you explain more about what we are doing here? Why its needed and why we always set the cache, why we change ->path and re-save it?

Reply:
here are 2 blocks to set alerts for a few pages, for the rest, there is 1 block.
So we run this build process, for some pages, up to 2 times, once per alerts-block placed on the page.
However, the CID is built on the $current_path, which means the same CID can hold information for multiple alert blocks.
That is why I need to get the data first, to not override cached data from other block builds.
I use $data->$path to not override other paths and later process that data to make early requests on html.html.twig.
It is redundant to set a property with name $path and then set the '#path' with the $path value, but it is very easy to iterate that on JS and differentiate multiple blocks at the same time.